U4-7316 (wip)

skttl opened the pull request:
http://issues.umbraco.org/issue/U4-7316

### Description
This PR adds a new property editor, allowing the editor to pick media images, and then on a per node basis, add crops to the picked images.

I have added a `Udi` property to the `ImageCropDataSet` model, and added additional methods for `GetCropUrl` using that `ImageCropDataSet`.

A new property editor, creating an array of `ImageCropDataSets` is added.

It works in the backoffice, when selecting media and adding crop info. But I'm currently stuck on getting the valueconverter to work. I get this message, when I start the solution:
`Type 'Umbraco.Core.PropertyEditors.ValueConverters.JsonValueConverter' cannot be an IPropertyValueConverter for property 'testMediaCrop' of content type 'home' because type 'Umbraco.Web.PropertyEditors.ValueConverters.MediaCropperPropertyConverter' has already been detected as a converter for that property, and only one converter can exist for a property.. This usually indicates that the content cache is corrupt; the content cache has been rebuilt in an attempt to self-fix the issue.`

I could use some help sorting that one out, and also some feedback on the direction I'm going.

https://i.imgur.com/0Gf30o6.gifv

I know the cropping dialog is buggy ATM, but I wan't to get the value converter working, before fixing the dialog.

leekelleher commented on src/Umbraco.Web/PropertyEditors/ValueConverters/MediaCropperPropertyConverter.cs:
```diff
@@ -0,0 +1,270 @@
+    /// <summary>
+    /// The media picker property value converter.
+    /// </summary>
+    [DefaultPropertyValueConverter]
```
@skttl Try doing what the `RelatedLinksPropertyConverter` value-converter does, by adding in the "shadow converter"
e.g. https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Web/PropertyEditors/ValueConverters/RelatedLinksEditorValueConvertor.cs#L28

`[DefaultPropertyValueConverter(typeof(JsonValueConverter))]`

I'm not entirely sure what "shadow" means in this context, but if the related-links converter is doing it, then it's worth a try?

skttl replied:
That worked! Thank you very much :)

ronaldbarendse replied:
The default `JsonValueConverter` is automatically registered for all editors that store their value as JSON, but because the `MediaCropperPropertyConverter` is also a default converter, Umbraco doesn't know which one to use...

When writing custom (non-default) PVCers, you don't have to do this BTW, since they always override (shadow) default converters 👍

skttl commented:
Got the value converter working.

![image](https://user-images.githubusercontent.com/3726467/45027519-e9ff8500-b041-11e8-8df4-2261d345e36a.png)


```
@foreach (var media in Model.Content.TestMediaCrop)
{ 
    <img src="@media.GetCropUrl("testcrop")" /><br />
    <img src="@media.GetCropUrl("testcrop2")" /><br />
    <img src="@media.GetCropUrl(100,100)" />
    <hr />
}
```

Next up, fix up the cropper dialog :)

skttl commented:
Added some UI fixes, so the custom crops fit better in the dialog. Also changed the close button on the "custom crop"-view, to a real `button` in order to not close the dialog when the user clicks it.

I also made it reuse any existing matching custom crops, or focalPoints upon selecting a media item.

It's all showed in the video: https://streamable.com/2cm88

ronaldbarendse commented on src/Umbraco.Web/ImageCropperTemplateExtensions.cs:
```diff
@@ -239,6 +239,95 @@ public static string GetCropUrl(this IPublishedContent mediaItem, string propert
+        {
+            if (cropDataSet.Udi == null) throw new ArgumentNullException("cropDataSet.Udi");
+
+            var mediaId = ApplicationContext.Current.Services.EntityService.GetIdForUdi(cropDataSet.Udi);
```
I'm not sure using both the `ApplicationContext.Current` and `UmbracoContext.Current` singletons and the `EntityService` (although cached) are desirable here...

skttl commented:
@ronaldbarendse thanks for your great feedback. It makes good sense to me, I'll refactor :)
